Rework ArgumentCodersCocoa to get us much closer to generated serializers

beidson opened the pull request:
#### f10ee49ef636c10dd181239c2b624bf85a954ebc
<pre>
Rework ArgumentCodersCocoa to get us much closer to generated serializers
<a href="https://bugs.webkit.org/show_bug.cgi?id=264588">https://bugs.webkit.org/show_bug.cgi?id=264588</a>
<a href="https://rdar.apple.com/118236121">rdar://118236121</a>

Reviewed by Alex Christensen.

This changes introduces ObjC object wrappers.
Encoding an ObjC object directly creates a typed wrapper around it.

For wrappers around a type that we still do directly via ObjC (e.g. NSArray), encoding the wrapper
then ends up going down the old &quot;encode ObjC directly&quot; path.

This change has the following benefits:
1 - We now broadly use serialization.in files for all ObjC types.
2 - For catch-all types like SecureCoding, we can now easily break out individual types into
    their own individual wrappers with individual serialization.
3 - So it follows, over time, the complicated logic in ArgumentCodersCocoa.mm can be whittled away.
4 - Having wrappers enables us to directly serialize collection types (like NSArray) without
    relying on the existing ObjC code path

* Source/WebCore/platform/mac/SerializedPlatformDataCueMac.h:
* Source/WebCore/platform/mac/SerializedPlatformDataCueMac.mm:
(WebCore::SerializedPlatformDataCueMac::allowedClassesForNativeValues):
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Platform/IPC/Decoder.h:
(IPC::Decoder::decodeWithAllowedClasses):
(IPC::Decoder::allowedClasses const):
* Source/WebKit/Scripts/generate-serializers.py:
(decode_type):
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.h:
(IPC::CoreIPCRetainPtr::CoreIPCRetainPtr):
(IPC::isObjectClassAllowed):
(IPC::decodeRequiringAllowedClasses):
(IPC::ArgumentCoder&lt;T::encode):
(IPC::ArgumentCoder&lt;CoreIPCRetainPtr&lt;T&gt;&gt;::encode):
(IPC::ArgumentCoder&lt;CoreIPCRetainPtr&lt;T&gt;&gt;::decode):
(IPC::decodeWithAllowedClasses): Deleted.
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm:
(IPC::typeFromObject):
(IPC::encodeObjectDirectly&lt;NSArray&gt;):
(IPC::decodeObjectDirectlyRequiringAllowedClasses&lt;NSArray&gt;):
(IPC::encodeObjectDirectly&lt;NSDictionary&gt;):
(IPC::id&gt;&gt;):
(IPC::decodeObjectDirectlyRequiringAllowedClasses&lt;NSDictionary&gt;):
(IPC::encodeObjectDirectly&lt;WebCore::CocoaFont&gt;):
(IPC::decodeObjectDirectlyRequiringAllowedClasses&lt;WebCore::CocoaFont&gt;):
(IPC::encodeObjectDirectly&lt;NSObject&lt;NSSecureCoding&gt;&gt;):
(IPC::shouldEnableStrictMode):
(IPC::decodeObjectDirectlyRequiringAllowedClasses&lt;NSObject&lt;NSSecureCoding&gt;&gt;):
(IPC::encodeObjectDirectly&lt;CFTypeRef&gt;):
(IPC::decodeObjectDirectlyRequiringAllowedClasses&lt;CFTypeRef&gt;):
(IPC::encodeObjectWithWrapper):
(IPC::decodeObjectFromWrapper):
(IPC::encodeArrayInternal): Deleted.
(IPC::decodeArrayInternal): Deleted.
(IPC::encodeColorInternal): Deleted.
(IPC::decodeColorInternal): Deleted.
(IPC::encodeDataInternal): Deleted.
(IPC::decodeDataInternal): Deleted.
(IPC::encodeDateInternal): Deleted.
(IPC::decodeDateInternal): Deleted.
(IPC::encodeDictionaryInternal): Deleted.
(IPC::decodeDictionaryInternal): Deleted.
(IPC::encodeFontInternal): Deleted.
(IPC::decodeFontInternal): Deleted.
(IPC::encodeNumberInternal): Deleted.
(IPC::decodeNumberInternal): Deleted.
(IPC::encodeSecureCodingInternal): Deleted.
(IPC::decodeSecureCodingInternal): Deleted.
(IPC::encodeStringInternal): Deleted.
(IPC::decodeStringInternal): Deleted.
(IPC::encodeURLInternal): Deleted.
(IPC::decodeURLInternal): Deleted.
(IPC::encodeCFInternal): Deleted.
(IPC::decodeCFInternal): Deleted.
(IPC::encodeObject): Deleted.
(IPC::decodeObject): Deleted.
* Source/WebKit/Shared/Cocoa/CoreIPCArray.h: Copied from Source/WebKit/Shared/Cocoa/CoreIPCDate.h.
(WebKit::CoreIPCArray::CoreIPCArray):
(WebKit::CoreIPCArray::toID):
* Source/WebKit/Shared/Cocoa/CoreIPCArray.serialization.in: Added.
* Source/WebKit/Shared/Cocoa/CoreIPCCFType.h: Copied from Source/WebKit/Shared/Cocoa/CoreIPCDate.h.
(WebKit::CoreIPCCFType::CoreIPCCFType):
(WebKit::CoreIPCCFType::toID):
* Source/WebKit/Shared/Cocoa/CoreIPCCFType.serialization.in: Added.
* Source/WebKit/Shared/Cocoa/CoreIPCColor.h: Copied from Source/WebKit/Shared/Cocoa/CoreIPCDate.h.
(WebKit::CoreIPCColor::CoreIPCColor):
(WebKit::CoreIPCColor::toID):
* Source/WebKit/Shared/Cocoa/CoreIPCColor.serialization.in: Added.
* Source/WebKit/Shared/Cocoa/CoreIPCData.h:
(WebKit::CoreIPCData::CoreIPCData):
(WebKit::CoreIPCData::toID const):
* Source/WebKit/Shared/Cocoa/CoreIPCDate.h:
(WebKit::CoreIPCDate::CoreIPCDate):
(WebKit::CoreIPCDate::toID const):
* Source/WebKit/Shared/Cocoa/CoreIPCDictionary.h: Copied from Source/WebKit/Shared/Cocoa/CoreIPCDate.h.
(WebKit::CoreIPCDictionary::CoreIPCDictionary):
(WebKit::CoreIPCDictionary::toID):
* Source/WebKit/Shared/Cocoa/CoreIPCDictionary.serialization.in: Added.
* Source/WebKit/Shared/Cocoa/CoreIPCFont.h: Copied from Source/WebKit/Shared/Cocoa/CoreIPCDate.h.
(WebKit::CoreIPCFont::CoreIPCFont):
(WebKit::CoreIPCFont::toID):
* Source/WebKit/Shared/Cocoa/CoreIPCFont.serialization.in: Added.
* Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.h: Copied from Source/WebKit/Shared/Cocoa/CoreIPCData.h.
(WebKit::CoreIPCNSCFObject::CoreIPCNSCFObject):
* Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.mm: Added.
(WebKit::valueFromID):
(WebKit::CoreIPCNSCFObject::CoreIPCNSCFObject):
(WebKit::CoreIPCNSCFObject::toID):
* Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.serialization.in: Added.
* Source/WebKit/Shared/Cocoa/CoreIPCSecureCoding.h: Copied from Source/WebKit/Shared/Cocoa/CoreIPCDate.h.
(WebKit::CoreIPCSecureCoding::toID):
* Source/WebKit/Shared/Cocoa/CoreIPCSecureCoding.mm: Copied from Source/WebKit/Shared/Cocoa/CoreIPCDate.h.
(WebKit::CoreIPCSecureCoding::CoreIPCSecureCoding):
* Source/WebKit/Shared/Cocoa/CoreIPCSecureCoding.serialization.in: Added.
* Source/WebKit/Shared/Cocoa/CoreIPCString.h: Copied from Source/WebKit/Shared/Cocoa/CoreIPCDate.h.
(WebKit::CoreIPCString::CoreIPCString):
(WebKit::CoreIPCString::toID):
* Source/WebKit/Shared/Cocoa/CoreIPCString.serialization.in: Added.
* Source/WebKit/Shared/Cocoa/CoreIPCURL.h: Copied from Source/WebKit/Shared/Cocoa/CoreIPCDate.h.
(WebKit::CoreIPCURL::CoreIPCURL):
(WebKit::CoreIPCURL::toID):
* Source/WebKit/Shared/Cocoa/CoreIPCURL.serialization.in: Added.
* Source/WebKit/Shared/Cocoa/WebIconUtilities.mm:
* Source/WebKit/Shared/cf/CoreIPCNumber.h:
(WebKit::CoreIPCNumber::CoreIPCNumber):
(WebKit::CoreIPCNumber::toID):
* Source/WebKit/Shared/mac/WebCoreArgumentCodersMac.mm:
(IPC::ArgumentCoder&lt;WebCore::SerializedPlatformDataCueValue&gt;::encodePlatformData):
(IPC::ArgumentCoder&lt;WebCore::SerializedPlatformDataCueValue&gt;::decodePlatformData):
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:
(TEST):

Canonical link: <a href="https://commits.webkit.org/270587@main">https://commits.webkit.org/270587@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0fe74be7745d6a7610495fa636417dfae964deb0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/25851 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4458 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27128 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/27948 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23667 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/26166 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6218 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/1885 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/23760 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26100 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/3353 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22269 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/28528 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/26013 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/2969 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/23224 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/29289 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/23586 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/23604 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27163 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/2996 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1217 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4387 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6213 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3452 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3312 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/Platform/IPC/Decoder.h:
```diff
@@ -185,6 +202,8 @@ class Decoder {
     ImportanceAssertion m_importanceAssertion;
 #endif
 
+    Vector<ClassStructPtr> m_allowedClasses;
```
This could be inside PLATFORM(COCOA) since no other platforms decode ObjC types.

achristensen07 commented on Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.h:
```diff
@@ -32,16 +34,56 @@
+    }
+};
+
+enum class NSType {
```
: uint8_t

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.mm:
```diff
@@ -0,0 +1,96 @@
+};
+
+template<class... Ts>
+overload(Ts...) -> overload<Ts...>;
```
makeVisitor?  switchOn?

beidson replied:
I...  Didn't know about those. Obviously way better.

achristensen07 commented on Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm:
```diff
@@ -719,10 +648,12 @@ static bool shouldEnableStrictMode(Decoder& decoder, NSArray<Class> *allowedClas
     unarchiver.get().delegate = delegate.get();
 
-    auto allowedClassSet = adoptNS([[NSMutableSet alloc] initWithArray:allowedClasses]);
+    auto allowedClassSet = adoptNS([[NSMutableSet alloc] init]);
```
initWithCapacity would be slightly better here.

achristensen07 commented on Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm:
```diff
@@ -719,10 +648,12 @@ static bool shouldEnableStrictMode(Decoder& decoder, NSArray<Class> *allowedClas
-    auto allowedClassSet = adoptNS([[NSMutableSet alloc] initWithArray:allowedClasses]);
+    auto allowedClassSet = adoptNS([[NSMutableSet alloc] init]);
+    for (auto allowedClass : allowedClasses)
+        [allowedClassSet addObject:(allowedClass)];
```
parentheses not needed.

achristensen07 commented on Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm:
```diff
@@ -669,45 +596,47 @@ static bool shouldEnableStrictMode(Decoder& decoder, NSArray<Class> *allowedClas
     // If you want to serialize something new, extract its contents into a
     // struct and use a *.serialization.in file to serialize its contents.
-    RELEASE_LOG_FAULT(SecureCoding, "Strict mode check found unknown classes %@", allowedClasses);
+    RetainPtr<NSMutableArray> nsAllowedClasses = adoptNS([[NSMutableArray alloc] init]);
```
nit: initWithCapacity

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCCFType.h:
```diff
@@ -0,0 +1,56 @@
+    {
+    }
+
+    RetainPtr<id> toID() { return static_cast<id>(m_cfType.get()); }
```
(__bridge id) or bridge_cast?

beidson replied:
I was going for identical code to `decodeCFInternal` here.

But agreed that's an improvement.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCDictionary.serialization.in:
```diff
@@ -0,0 +1,31 @@
+
+#if PLATFORM(COCOA)
+
+webkit_platform_headers: "CoreIPCDictionary.h"
```
I bet we could just change the python to add to the correct set of headers based on the presence of the [WebKitPlatform] attribute rather than include all these explicitly.

beidson replied:
Yah, sounds reasonable.
Will do a targeted followup.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.h:
```diff
@@ -0,0 +1,79 @@
+        CoreIPCSecureCoding,
+        CoreIPCString,
+        CoreIPCURL,
+        std::nullptr_t
```
This should probably be first in case the std::variant default constructor is called anywhere.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.mm:
```diff
@@ -0,0 +1,89 @@
+    WTF::switchOn(m_value, [&](auto& object) {
+        result = object.toID();
+    }, [](std::nullptr_t) {
+        // result should be nil, which is the default value initialized above.
```
ASSERT_NOT_REACHED()?

beidson replied:
The existing serialization code can actually encode/decode nil successfully, which this handles.

`void encodeObject(Encoder& encoder, id object)
 {
    encoder << static_cast<bool>(!object);
    if (!object)
        return;`
        
 And:
 
 `std::optional<RetainPtr<id>> decodeObject(Decoder& decoder, NSArray<Class> *allowedClasses)
 {
    bool isNull;
    if (!decoder.decode(isNull))
        return std::nullopt;
    if (isNull)
        return { nullptr };`

achristensen07 commented on Source/WebKit/WebKit.xcodeproj/project.pbxproj:
```diff
@@ -3,7 +3,7 @@
 	classes = {
 	};
-	objectVersion = 55;
+	objectVersion = 54;
```
Undo this change.

beidson committed:
Rework ArgumentCodersCocoa to get us much closer to generated serializers
https://bugs.webkit.org/show_bug.cgi?id=264588
rdar://118236121

Reviewed by Alex Christensen.

This changes introduces ObjC object wrappers.
Encoding an ObjC object directly creates a typed wrapper around it.

For wrappers around a type that we still do directly via ObjC (e.g. NSArray), encoding the wrapper
then ends up going down the old "encode ObjC directly" path.

This change has the following benefits:
1 - We now broadly use serialization.in files for all ObjC types.
2 - For catch-all types like SecureCoding, we can now easily break out individual types into
    their own individual wrappers with individual serialization.
3 - So it follows, over time, the complicated logic in ArgumentCodersCocoa.mm can be whittled away.
4 - Having wrappers enables us to directly serialize collection types (like NSArray) without
    relying on the existing ObjC code path

* Source/WebCore/platform/mac/SerializedPlatformDataCueMac.h:
* Source/WebCore/platform/mac/SerializedPlatformDataCueMac.mm:
(WebCore::SerializedPlatformDataCueMac::allowedClassesForNativeValues):
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Platform/IPC/Decoder.h:
(IPC::Decoder::decodeWithAllowedClasses):
(IPC::Decoder::allowedClasses const):
* Source/WebKit/Scripts/generate-serializers.py:
(decode_type):
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.h:
(IPC::CoreIPCRetainPtr::CoreIPCRetainPtr):
(IPC::isObjectClassAllowed):
(IPC::decodeRequiringAllowedClasses):
(IPC::ArgumentCoder<T::encode):
(IPC::ArgumentCoder<CoreIPCRetainPtr<T>>::encode):
(IPC::ArgumentCoder<CoreIPCRetainPtr<T>>::decode):
(IPC::decodeWithAllowedClasses): Deleted.
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm:
(IPC::typeFromObject):
(IPC::encodeObjectDirectly<NSArray>):
(IPC::decodeObjectDirectlyRequiringAllowedClasses<NSArray>):
(IPC::encodeObjectDirectly<NSDictionary>):
(IPC::id>>):
(IPC::decodeObjectDirectlyRequiringAllowedClasses<NSDictionary>):
(IPC::encodeObjectDirectly<WebCore::CocoaFont>):
(IPC::decodeObjectDirectlyRequiringAllowedClasses<WebCore::CocoaFont>):
(IPC::encodeObjectDirectly<NSObject<NSSecureCoding>>):
(IPC::shouldEnableStrictMode):
(IPC::decodeObjectDirectlyRequiringAllowedClasses<NSObject<NSSecureCoding>>):
(IPC::encodeObjectDirectly<CFTypeRef>):
(IPC::decodeObjectDirectlyRequiringAllowedClasses<CFTypeRef>):
(IPC::encodeObjectWithWrapper):
(IPC::decodeObjectFromWrapper):
(IPC::encodeArrayInternal): Deleted.
(IPC::decodeArrayInternal): Deleted.
(IPC::encodeColorInternal): Deleted.
(IPC::decodeColorInternal): Deleted.
(IPC::encodeDataInternal): Deleted.
(IPC::decodeDataInternal): Deleted.
(IPC::encodeDateInternal): Deleted.
(IPC::decodeDateInternal): Deleted.
(IPC::encodeDictionaryInternal): Deleted.
(IPC::decodeDictionaryInternal): Deleted.
(IPC::encodeFontInternal): Deleted.
(IPC::decodeFontInternal): Deleted.
(IPC::encodeNumberInternal): Deleted.
(IPC::decodeNumberInternal): Deleted.
(IPC::encodeSecureCodingInternal): Deleted.
(IPC::decodeSecureCodingInternal): Deleted.
(IPC::encodeStringInternal): Deleted.
(IPC::decodeStringInternal): Deleted.
(IPC::encodeURLInternal): Deleted.
(IPC::decodeURLInternal): Deleted.
(IPC::encodeCFInternal): Deleted.
(IPC::decodeCFInternal): Deleted.
(IPC::encodeObject): Deleted.
(IPC::decodeObject): Deleted.
* Source/WebKit/Shared/Cocoa/CoreIPCArray.h: Copied from Source/WebKit/Shared/Cocoa/CoreIPCDate.h.
(WebKit::CoreIPCArray::CoreIPCArray):
(WebKit::CoreIPCArray::toID):
* Source/WebKit/Shared/Cocoa/CoreIPCArray.serialization.in: Added.
* Source/WebKit/Shared/Cocoa/CoreIPCCFType.h: Copied from Source/WebKit/Shared/Cocoa/CoreIPCDate.h.
(WebKit::CoreIPCCFType::CoreIPCCFType):
(WebKit::CoreIPCCFType::toID):
* Source/WebKit/Shared/Cocoa/CoreIPCCFType.serialization.in: Added.
* Source/WebKit/Shared/Cocoa/CoreIPCColor.h: Copied from Source/WebKit/Shared/Cocoa/CoreIPCDate.h.
(WebKit::CoreIPCColor::CoreIPCColor):
(WebKit::CoreIPCColor::toID):
* Source/WebKit/Shared/Cocoa/CoreIPCColor.serialization.in: Added.
* Source/WebKit/Shared/Cocoa/CoreIPCData.h:
(WebKit::CoreIPCData::CoreIPCData):
(WebKit::CoreIPCData::toID const):
* Source/WebKit/Shared/Cocoa/CoreIPCDate.h:
(WebKit::CoreIPCDate::CoreIPCDate):
(WebKit::CoreIPCDate::toID const):
* Source/WebKit/Shared/Cocoa/CoreIPCDictionary.h: Copied from Source/WebKit/Shared/Cocoa/CoreIPCDate.h.
(WebKit::CoreIPCDictionary::CoreIPCDictionary):
(WebKit::CoreIPCDictionary::toID):
* Source/WebKit/Shared/Cocoa/CoreIPCDictionary.serialization.in: Added.
* Source/WebKit/Shared/Cocoa/CoreIPCFont.h: Copied from Source/WebKit/Shared/Cocoa/CoreIPCDate.h.
(WebKit::CoreIPCFont::CoreIPCFont):
(WebKit::CoreIPCFont::toID):
* Source/WebKit/Shared/Cocoa/CoreIPCFont.serialization.in: Added.
* Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.h: Copied from Source/WebKit/Shared/Cocoa/CoreIPCData.h.
(WebKit::CoreIPCNSCFObject::CoreIPCNSCFObject):
* Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.mm: Added.
(WebKit::valueFromID):
(WebKit::CoreIPCNSCFObject::CoreIPCNSCFObject):
(WebKit::CoreIPCNSCFObject::toID):
* Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.serialization.in: Added.
* Source/WebKit/Shared/Cocoa/CoreIPCSecureCoding.h: Copied from Source/WebKit/Shared/Cocoa/CoreIPCDate.h.
(WebKit::CoreIPCSecureCoding::toID):
* Source/WebKit/Shared/Cocoa/CoreIPCSecureCoding.mm: Copied from Source/WebKit/Shared/Cocoa/CoreIPCDate.h.
(WebKit::CoreIPCSecureCoding::CoreIPCSecureCoding):
* Source/WebKit/Shared/Cocoa/CoreIPCSecureCoding.serialization.in: Added.
* Source/WebKit/Shared/Cocoa/CoreIPCString.h: Copied from Source/WebKit/Shared/Cocoa/CoreIPCDate.h.
(WebKit::CoreIPCString::CoreIPCString):
(WebKit::CoreIPCString::toID):
* Source/WebKit/Shared/Cocoa/CoreIPCString.serialization.in: Added.
* Source/WebKit/Shared/Cocoa/CoreIPCURL.h: Copied from Source/WebKit/Shared/Cocoa/CoreIPCDate.h.
(WebKit::CoreIPCURL::CoreIPCURL):
(WebKit::CoreIPCURL::toID):
* Source/WebKit/Shared/Cocoa/CoreIPCURL.serialization.in: Added.
* Source/WebKit/Shared/Cocoa/WebIconUtilities.mm:
* Source/WebKit/Shared/cf/CoreIPCNumber.h:
(WebKit::CoreIPCNumber::CoreIPCNumber):
(WebKit::CoreIPCNumber::toID):
* Source/WebKit/Shared/mac/WebCoreArgumentCodersMac.mm:
(IPC::ArgumentCoder<WebCore::SerializedPlatformDataCueValue>::encodePlatformData):
(IPC::ArgumentCoder<WebCore::SerializedPlatformDataCueValue>::decodePlatformData):
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:
(TEST):

Canonical link: https://commits.webkit.org/270587@main
